### PR TITLE
Added reboot azure stress test and fixed scripts

### PR DESCRIPTION
**Pull request by mbivolan**
1. Added azure reboot stress test
2. Fixed xml save failure when a relative path is given as a parameter, for the following scripts:
RunTests.ps1
UpdateGlobalConfigurationFromXmlSecrets.ps1

### CODE REVIEW

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+    for ( $rebootNr = 1; $rebootNr -le $rebootNumber; $rebootNr++ ) {
+        try
+        {       
+            $hs1VIP = $AllVMData.PublicIP
+            $hs1vm1sshport = $AllVMData.SSHPort
+            $hs1ServiceUrl = $AllVMData.URL
+            $hs1vm1Dip = $AllVMData.InternalIP
+            LogMsg "Trying to restart $($AllVMData.RoleName): $rebootNr / $rebootNumber ..."
+            if ( $UseAzureResourceManager )
```
Comment (lizzha) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: This condition is not necessary, as ASM is not supported in LISAv2.

File: XML/TestCases/StressTests.xml [hunk 2]
```diff
@@ -1,3 +1,17 @@
 <TestCases>
-
+    <test>
+        <testName>STRESSTEST-VERIFY-RESTART</testName>
+        <testScript></testScript>
+        <PowershellScript>STRESSTEST-VERIFY-RESTART.ps1</PowershellScript>
+        <files></files>
+        <setupType>SingleVM</setupType>
+        <TestParameters>
+            <param>rebootNumber=2</param>
```
Comment (lizzha) [inline XML/TestCases/StressTests.xml]: Just want to make sure the reboot number is expected

Comment (mbivolan) [reply XML/TestCases/StressTests.xml]: You are right. Here the default value should be 1.I amended the first commit with the changes requested.

File: XML/TestCases/StressTests.xml [hunk 3]
```diff
@@ -1,3 +1,17 @@
...
+    <test>
+        <testName>STRESSTEST-VERIFY-RESTART</testName>
+        <testScript></testScript>
+        <PowershellScript>STRESSTEST-VERIFY-RESTART.ps1</PowershellScript>
+        <files></files>
+        <setupType>SingleVM</setupType>
+        <TestParameters>
+            <param>rebootNumber=1</param>
+        </TestParameters>
+        <Platform>Azure</Platform>
```
Comment (juhlee-microsoft) [inline XML/TestCases/StressTests.xml]: Azure & Hyper-V both?

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    LogMsg "Reboot Stress Test Result: $rebootNr/$rebootNumber"
+    if ( ($rebootNr - 1) -lt $rebootNumber )
+    {
+        $testResult = "FAIL"
+    }
+}
+else
+{
+	$testResult = "FAIL"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: It is nice to print out the failure reason in the log?

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        try
+        {       
+            $hs1VIP = $AllVMData.PublicIP
+            $hs1vm1sshport = $AllVMData.SSHPort
+            $hs1ServiceUrl = $AllVMData.URL
+            $hs1vm1Dip = $AllVMData.InternalIP
+            LogMsg "Trying to restart $($AllVMData.RoleName): $rebootNr / $rebootNumber ..."
+            Restart-AzureRmVM -ResourceGroupName $AllVMData.ResourceGroupName -Name $AllVMData.RoleName -Verbose
+            $restartVM = $?
+            if ( $restartVM )
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: This should be only condition - $restartVM.Status -eq "Succeeded"

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                $testResult = "FAIL"
+                break
+            }
+
+        }
+        catch
+        {
+            $ErrorMessage =  $_.Exception.Message
+            LogMsg "EXCEPTION : $ErrorMessage"   
+            break
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: this 'break' statement cause unassigned value of $testResult outside. Please check again.

File: XML/TestCases/StressTests.xml (hunk 3 above)
Comment (chvalean) [reply XML/TestCases/StressTests.xml]: This test case is an addition to the developer pipeline use, in order to support Azure provisioning and reboot loop.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 6 above)
Comment (mbivolan) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: The script execution will continue with the "Finally" block (even if the "break" command exists) and the proper value will be assigned to the $testResult variable.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 5 above)
Comment (mbivolan) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: The $restartVM variable contains a bool variable, it has no "Status" attribute.
If you meant that the $restartVM should contain the return value from the "Restart-AzureRmVM" command, I think it's a bug there, as all the attributes are empty.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 4 above)
Comment (mbivolan) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: done

File: Utilities/UpdateGlobalConfigurationFromXmlSecrets.ps1 [hunk 7]
```diff
@@ -10,8 +10,10 @@
 param(
```
Comment (ader1990) [inline Utilities/UpdateGlobalConfigurationFromXmlSecrets.ps1]: this change should go in a separate commit

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 8]
```diff
@@ -0,0 +1,96 @@
+$result = ""
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: always use the PowerShell guidelines (execution policy, main script, methods, consistent indentation and use of {)

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+$testResult = ""
+$resultArr = @()
+
+$isDeployed = DeployVMS -setupType $currentTestData.setupType -Distro $Distro -xmlConfig $xmlConfig
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: line too long

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+if ($isDeployed)
+{
+    foreach ( $param in $currentTestData.TestParameters.param )
+	{
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: you are waisting one line of code each time you do this

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        $rebootNumber = "1"
+    }
+    for ( $rebootNr = 1; $rebootNr -le $rebootNumber; $rebootNr++ ) {
+        try
+        {       
+            $hs1VIP = $AllVMData.PublicIP
+            $hs1vm1sshport = $AllVMData.SSHPort
+            $hs1ServiceUrl = $AllVMData.URL
+            $hs1vm1Dip = $AllVMData.InternalIP
+            LogMsg "Trying to restart $($AllVMData.RoleName): $rebootNr / $rebootNumber ..."
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: use -f

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+            Restart-AzureRmVM -ResourceGroupName $AllVMData.ResourceGroupName -Name $AllVMData.RoleName -Verbose
+            $restartVM = $?
+            if ( $restartVM )
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: parentheses incorrect  usage

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+{
+	$testResult = "FAIL"
+	LogErr "The current deployment could not be found"
+	$resultArr += $testResult
+}
+
+$result = GetFinalResultHeader -resultarr $resultArr
+
+#Clean up the setup
+DoTestCleanUp -result $result -testName $currentTestData.testName -deployedServices $isDeployed -ResourceGroups $isDeployed
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: line too long

Comment (ader1990) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: where are this variables coming from?
if you happen to somehow use global variables, declare them in a comment at the beginning of the script and send them all as params to the Main function (which you need to create)

Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: and where are these methods coming from? there is no module loaded or dot sourced .ps1
this is a serious issue when trying to make sense of this code

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 (hunk 10 above)
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: never use tabs and spaces in the same file, for indentation, always go for spaces as any editor shows them correctly. tabs are not shown nicely in other editors like vim

Comment (ader1990) [reply Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: You are literally wasting 20 lines of code by doing this in the current file.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+$ErrorActionPreference = "Continue"
+
+function Main {
+    param(
+        [parameter(Mandatory=$true)]
+        [String] $TestData,
+        [parameter(Mandatory=$true)]
+        [String] $Distro,
+        [parameter(Mandatory=$true)]
+        [String] $xmlConfig,
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: always use upper caps for param names

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                LogMsg ("Trying to restart {0}: {1} / {2} ..." `
+                    -f @($VMData.RoleName, $rebootNr, $rebootNumber))
+                Restart-AzureRmVM -ResourceGroupName $VMData.ResourceGroupName `
+                    -Name $VMData.RoleName -Verbose
+                $restartVM = $?
+                if ($restartVM) {
+                    $isSSHOpened = isAllSSHPortsEnabledRG -VMDataObject $VMData
+                    if ($isSSHOpened -eq "True") {
+                        $isRestarted = $true
+                    }
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: put else on this line, same goes with catch, finally, etc

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                    $testResult = "FAIL"
+                    break
+                }
+            }
+            catch {
+                $ErrorMessage =  $_.Exception.Message
+                LogMsg "EXCEPTION : $ErrorMessage"
+                break
+            }
+            Finally {
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: be consistent and use either upper case or lowercase for keywords.
in this case, to be consistent, use lower case.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    else {
+        $testResult = "FAIL"
+        LogErr "The current deployment could not be found"
+        $resultArr += $testResult
+    }
+
+    $result = GetFinalResultHeader -resultarr $resultArr
+
+    #Clean up the setup
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: leave a space between # and the rest of the message

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    DoTestCleanUp -result $result -testName $TestData.testName `
+        -deployedServices $isDeployed -ResourceGroups $isDeployed
+
+    #Return the result and summery to the test suite script..
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: ditto
small typo: summary

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 19]
```diff
@@ -0,0 +1,107 @@
+# Global Variables
```
Comment (ader1990) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: put this section before the Main call at the end

File: XML/TestCases/StressTests.xml (hunk 3 above)
Comment (juhlee-microsoft) [reply XML/TestCases/StressTests.xml]: acked. Then leave it as 'Azure'

**Comment by juhlee-microsoft** [comment]
As soon as @ader1990 signs off, this PR can merge to the master.

File: XML/TestCases/StressTests.xml [hunk 20]
```diff
@@ -1,3 +1,17 @@
 <TestCases>
-
+    <test>
+        <testName>STRESSTEST-VERIFY-RESTART</testName>
+        <testScript></testScript>
+        <PowershellScript>STRESSTEST-VERIFY-RESTART.ps1</PowershellScript>
+        <files></files>
+        <setupType>SingleVM</setupType>
+        <TestParameters>
+            <param>rebootNumber=1</param>
```
Comment (iamshital) [inline XML/TestCases/StressTests.xml]: rebootNumber = 1 results in exact duplicate test.
There is already a test which deploys VM -> Reboots it once and passes the test. 
As this is supposed to be boot stress test, add at least 100 iterations.

File: Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1 [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        for ($rebootNr = 1; $rebootNr -le $rebootNumber; $rebootNr++) {
+            try {
+                $hs1VIP = $AllVMData.PublicIP
+                $hs1vm1sshport = $AllVMData.SSHPort
+                $hs1ServiceUrl = $AllVMData.URL
+                $hs1vm1Dip = $AllVMData.InternalIP
+                LogMsg ("Trying to restart {0}: {1} / {2} ..." `
+                    -f @($AllVMData.RoleName, $rebootNr, $rebootNumber))
+                Restart-AzureRmVM -ResourceGroupName $AllVMData.ResourceGroupName `
```
Comment (iamshital) [inline Testscripts/Windows/STRESSTEST-VERIFY-RESTART.ps1]: Use predefined method :
`$RestartStatus = RestartAllDeployments -allVMData $allVMData`
`if($RestartStatus -eq "True")`
`{`
`   // Your statememets`
`}`
... to make the test case independent of the platform.

File: XML/TestCases/StressTests.xml (hunk 3 above)
Comment (iamshital) [reply XML/TestCases/StressTests.xml]: @chvalean , if you want to use single deployment and reboot scenario for your developer pipeline - use "VERIFY-DEPLOYMENT-PROVISION" test.
